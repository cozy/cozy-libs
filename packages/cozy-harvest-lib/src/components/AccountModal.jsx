import React, { Component } from 'react'
import PropTypes from 'prop-types'

import get from 'lodash/get'
import { withMutations } from 'cozy-client'
import { withRouter } from 'react-router'
import Spinner from 'cozy-ui/transpiled/react/Spinner'
import { ModalContent } from 'cozy-ui/transpiled/react/Modal'
import Infos from 'cozy-ui/transpiled/react/Infos'
import Button from 'cozy-ui/transpiled/react/Button'
import { translate } from 'cozy-ui/transpiled/react/I18n'

import accountMutations from '../connections/accounts'
import triggersMutations from '../connections/triggers'
import * as triggersModel from '../helpers/triggers'
import KonnectorConfiguration from './KonnectorConfiguration/KonnectorConfiguration'
import AccountSelectBox from './AccountSelectBox/AccountSelectBox'
import KonnectorModalHeader from './KonnectorModalHeader'

/**
 * AccountModal take an accountId and a list of accounts containing their
 * respecting triggers and display the selected account and the accounts linked
 * to this konnector
 *
 * You have to pass an array of accounts containing their associated trigger:
 * ```
 * accounts: [
 *  {
 *    account: {
 *       _id: '',
 *    },
 *    trigger: {
 *      id: '',
 *      current_state: {
 *      }
 *    }
 *  }
 * ]
 * ```
 *
 * from this array and from the passed accountId, we can fetch the account.
 *
 */
class AccountModal extends Component {
  state = {
    trigger: null,
    account: null,
    fetching: true,
    error: false
  }
  componentDidMount() {
    this.loadSelectedAccountId()
  }

  componentDidUpdate(prevProps) {
    if (this.props.accountId !== prevProps.accountId)
      this.loadSelectedAccountId()
  }

  loadSelectedAccountId() {
    const { accountId, accounts } = this.props
    const matchingTrigger = get(
      accounts.find(account => account.account._id === accountId),
      'trigger'
    )
    if (matchingTrigger) this.fetchAccount(matchingTrigger)
  }

  async fetchAccount(trigger) {
    const { findAccount } = this.props
    this.setState({ fetching: true })

    try {
      const account = await findAccount(triggersModel.getAccountId(trigger))
      this.setState({
        account,
        trigger
      })
      return account
    } catch (error) {
      this.setState({
        error
      })
    } finally {
      this.setState({
        fetching: false
      })
    }
  }

  render() {
    const { konnector, onDismiss, history, accounts, t } = this.props
    const { trigger, account, fetching, error } = this.state
    return (
      <>
        <KonnectorModalHeader konnector={konnector}>
          {account && (
            <AccountSelectBox
              selectedAccount={account}
              accountsList={accounts}
              onChange={option => {
                history.push(`../${option.account._id}`)
              }}
              onCreate={() => {
                history.push(`../../new`)
              }}
            />
          )}
        </KonnectorModalHeader>
        <ModalContent>
          {error && (
            <Infos
              actionButton={
                <Button
                  theme="danger"
                  onClick={this.loadSelectedAccountId.bind(this)}
                  label={t('modal.konnector.error.button')}
                />
              }
              title={t('modal.konnector.error.title')}
              text={t('modal.konnector.error.description', error)}
              icon="warning"
              isImportant
            />
          )}
          {fetching && (
            <div className="u-ta-center">
              <Spinner size="xxlarge" />
            </div>
          )}
          {!error && !fetching && (
            <KonnectorConfiguration
              konnector={konnector}
              trigger={trigger}
              account={account}
              onAccountDeleted={onDismiss}
              addAccount={() => history.push('../new')}
            />
          )}
        </ModalContent>
      </>
    )
  }
}

AccountModal.propTypes = {
  konnector: PropTypes.object.isRequired,
  onDismiss: PropTypes.func,
  history: PropTypes.object.isRequired,
  accounts: PropTypes.array.isRequired,
  t: PropTypes.func.isRequired,
  findAccount: PropTypes.func.isRequired,
  accountId: PropTypes.string.isRequired
}
export default withMutations(accountMutations, triggersMutations)(
  withRouter(translate()(AccountModal))
)
