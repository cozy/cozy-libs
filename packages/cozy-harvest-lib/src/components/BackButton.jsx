import React from 'react'
import Icon from 'cozy-ui/transpiled/react/Icon'
import cx from 'classnames'
import withLocales from './hoc/withLocales'

const BackButton = props => {
  const { className, t, ...rest } = props

  return (
    <button
      className={cx(
        'u-inline-flex u-flex-items-center u-bg-white u-bdw-0 u-ph-0 u-pv-1 u-c-pointer',
        className
      )}
      {...rest}
    >
      <Icon icon="left" className="u-mr-half" />
      {t('Revenir Ã  la liste')}
    </button>
  )
}

export default withLocales(BackButton)
