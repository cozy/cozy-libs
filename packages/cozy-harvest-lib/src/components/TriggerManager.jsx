import React, { Component } from 'react'
import PropTypes from 'prop-types'

import { withMutations } from 'cozy-client'

import AccountForm from './AccountForm'
import OAuthForm from './OAuthForm'
import TwoFAForm from './TwoFAForm'
import accountsMutations from '../connections/accounts'
import { triggersMutations } from '../connections/triggers'
import filesMutations from '../connections/files'
import permissionsMutations from '../connections/permissions'
import accounts from '../helpers/accounts'
import cron from '../helpers/cron'
import konnectors from '../helpers/konnectors'
import triggers from '../helpers/triggers'
import withLocales from './hoc/withLocales'

const ERRORED = 'ERRORED'
const IDLE = 'IDLE'
const RUNNING = 'RUNNING'
const RUNNING_TWOFA = 'RUNNING_TWOFA'

const MODAL_PLACE_ID = 'coz-harvest-modal-place'

/**
 * Manage a trigger for a given konnector, from account edition to trigger
 * creation and launch.
 * @type {Component}
 */
export class TriggerManager extends Component {
  constructor(props) {
    super(props)
    const { account, trigger } = props

    this.handleNewAccount = this.handleNewAccount.bind(this)
    this.handleOAuthAccountId = this.handleOAuthAccountId.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
    this.closeTwoFAModal = this.closeTwoFAModal.bind(this)
    this.disableSuccessTimer = this.disableSuccessTimer.bind(this)
    this.handleLoginSuccessState = this.handleLoginSuccessState.bind(this)
    this.handleSuccess = this.handleSuccess.bind(this)
    this.handleError = this.handleError.bind(this)
    this.handleTwoFACodeAsked = this.handleTwoFACodeAsked.bind(this)
    this.handleSubmitTwoFACode = this.handleSubmitTwoFACode.bind(this)

    this.jobWatcher = null
    this.accountWatcher = null

    this.state = {
      account,
      error: triggers.getError(trigger),
      status: IDLE,
      trigger
    }
  }

  componentWillUnmount() {
    if (this.jobWatcher) this.jobWatcher.unsubscribeAll()
    if (this.accountWatcher) this.accountWatcher.unsubscribeAll()
  }

  closeTwoFAModal() {
    this.setState({
      status: RUNNING
    })
  }

  disableSuccessTimer() {
    if (this.jobWatcher) this.jobWatcher.disableSuccessTimer()
  }

  handleLoginSuccessState() {
    if (this.jobWatcher) this.jobWatcher.handleSuccess()
  }

  /**
   * Ensure that a trigger will exist, with valid destination folder with
   * permissions and references
   * @return {Object} Trigger document
   */
  async ensureTrigger() {
    const {
      addPermission,
      addReferencesTo,
      createDirectoryByPath,
      createTrigger,
      statDirectoryByPath,
      konnector,
      t
    } = this.props

    const { account, trigger } = this.state

    if (trigger) {
      return trigger
    }

    let folder

    if (konnectors.needsFolder(konnector)) {
      // TODO: remove previous legacy key `default.baseDir`
      // https://github.com/cozy/cozy-libs/issues/637
      const path = konnectors.buildFolderPath(konnector, account, {
        administrative: t('folder.administrative'),
        photos: t('folder.photos')
      })

      folder =
        (await statDirectoryByPath(path)) || (await createDirectoryByPath(path))

      await addPermission(konnector, konnectors.buildFolderPermission(folder))
      await addReferencesTo(konnector, [folder])
    }

    return await createTrigger(
      triggers.buildAttributes({
        account,
        cron: cron.fromKonnector(konnector),
        folder,
        konnector
      })
    )
  }

  /**
   * OAuth Form success handler. OAuthForm retrieves an account id created by the
   * cozy stack
   * @param  {string}  accountId
   */
  async handleOAuthAccountId(accountId) {
    const { findAccount } = this.props
    try {
      const oAuthAccount = await findAccount(accountId)
      return await this.handleNewAccount(oAuthAccount)
    } catch (error) {
      this.handleError(error)
    }
  }

  /**
   * Account creation success handler
   * @param  {Object}  account Created io.cozy.accounts document
   * @return {Object}          io.cozy.jobs document, runned with account data
   */
  async handleNewAccount(account) {
    this.setState({ account, error: null, status: RUNNING })
    const trigger = await this.ensureTrigger()
    this.setState({ trigger })
    this.accountWatcher = this.props.watchKonnectorAccount(account, {
      onTwoFACodeAsked: this.handleTwoFACodeAsked,
      onLoginSuccess: this.handleLoginSuccessState,
      onLoginSuccessHandled: this.disableSuccessTimer
    })
    return await this.launch(trigger)
  }

  handleError(error) {
    this.setState({
      error,
      status: ERRORED
    })

    const { onError } = this.props
    if (typeof onError === 'function') onError(error)
  }

  handleTwoFACodeAsked(statusCode) {
    const { status } = this.state
    if (accounts.isTwoFANeeded(status)) return
    // disable successTimeout since asked Two FA code
    this.disableSuccessTimer()
    this.setState({
      status: statusCode
    })
  }

  async handleSubmitTwoFACode(code) {
    const { konnector, saveAccount } = this.props
    const { account } = this.state
    this.setState({
      error: null,
      status: RUNNING_TWOFA
    })
    try {
      await saveAccount(konnector, accounts.updateTwoFaCode(account, code))
      if (this.jobWatcher) this.jobWatcher.enableSuccessTimer(10000)
    } catch (error) {
      return this.handleError(error)
    }
  }

  async handleSubmit(data = {}) {
    const { konnector, saveAccount } = this.props

    const { account } = this.state
    const isUpdate = !!account

    this.setState({
      error: null,
      status: RUNNING
    })

    try {
      const accountToSave = isUpdate
        ? accounts.mergeAuth(
            accounts.setSessionResetIfNecessary(
              accounts.resetState(account),
              data
            ),
            data
          )
        : accounts.build(konnector, data)
      const savedAccount = accounts.mergeAuth(
        await saveAccount(konnector, accountToSave),
        data
      )
      return await this.handleNewAccount(savedAccount)
    } catch (error) {
      return this.handleError(error)
    }
  }

  /**
   * Handle a success, typically job success or login success
   * @param  {Function} successCallback Typically onLoginSuccess or onSuccess
   * @param  {Array} args            Callback arguments
   */
  async handleSuccess(successCallback) {
    const trigger = await this.refetchTrigger()
    this.setState({ status: IDLE, trigger })
    if (typeof successCallback !== 'function') return
    successCallback(triggers)
  }

  /**
   * Launches a trigger
   * @param  {Object}  trigger io.cozy.triggers document
   * @return {Promise}         [description]
   */
  async launch(trigger) {
    const {
      launchTrigger,
      onLaunch,
      onLoginSuccess,
      onSuccess,
      watchKonnectorJob
    } = this.props
    this.jobWatcher = watchKonnectorJob(await launchTrigger(trigger))
    this.jobWatcher.on('error', this.handleError)
    this.jobWatcher.on('loginSuccess', () => this.handleSuccess(onLoginSuccess))
    this.jobWatcher.on('success', () => this.handleSuccess(onSuccess))

    if (typeof onLaunch === 'function') onLaunch(trigger)
  }

  async refetchTrigger() {
    const { fetchTrigger } = this.props
    const { trigger } = this.state
    try {
      return await fetchTrigger(trigger._id)
    } catch (error) {
      this.setState({ error, running: false })
      throw error
    }
  }

  render() {
    const { konnector, running, showError, modalContainerId } = this.props
    const { account, error, status } = this.state
    const submitting = status === RUNNING || running
    const submittingTwoFA = status === RUNNING_TWOFA
    const waitForTwoFACode = accounts.isTwoFANeeded(status)
    const isTwoFARetryCode = accounts.isTwoFARetry(status)
    const display2FA = waitForTwoFACode || submittingTwoFA || isTwoFARetryCode
    const modalInto = modalContainerId || MODAL_PLACE_ID

    const { oauth } = konnector

    if (oauth) {
      return (
        <OAuthForm
          account={account}
          konnector={konnector}
          onSuccess={this.handleOAuthAccountId}
          submitting={submitting || display2FA}
        />
      )
    }

    return (
      <div>
        <div id={modalInto} />
        <AccountForm
          account={account}
          error={error}
          konnector={konnector}
          onSubmit={this.handleSubmit}
          showError={showError}
          submitting={submitting || display2FA}
        />
        {display2FA && (
          <TwoFAForm
            account={account}
            konnector={konnector}
            dismissAction={this.closeTwoFAModal}
            handleSubmitTwoFACode={this.handleSubmitTwoFACode}
            submitting={submittingTwoFA}
            into={modalInto}
            retryAsked={isTwoFARetryCode && !submittingTwoFA}
          />
        )}
      </div>
    )
  }
}

TriggerManager.propTypes = {
  /**
   * Account document. Used to get intial form values.
   * If no account is passed, AccountForm will use empty initial values.
   * @type {Object}
   */
  account: PropTypes.object,
  /**
   * Konnector document. AccountForm will check the `fields` object to compute
   * fields.
   * @type {Object}
   */
  konnector: PropTypes.object.isRequired,
  /**
   * Indicates if the TriggerManager has to show errors. Sometimes errors may be
   * displayed elsewhere. However, a KonnectorJobError corresponding to a login
   * error is always displayed. Transmitted to AccountForm.
   * @type {Boolean}
   */
  showError: PropTypes.bool,
  /**
   * Existing trigger document to manage.
   * @type {Object}
   */
  trigger: PropTypes.object,
  /**
   * Indicates if the given trigger is already running, i.e. if it has been
   * launched and if an associated job with status 'running' exists.
   * @type {[type]}
   */
  running: PropTypes.bool,
  /**
   * Translation function
   */
  t: PropTypes.func,
  //
  // mutations
  //
  /**
   * Permission mutation
   * @type {Function}
   */
  addPermission: PropTypes.func,
  /**
   * File mutation
   * @type {Function}
   */
  addReferencesTo: PropTypes.func,
  /**
   * Trigger mutation
   * @type {Function}
   */
  createTrigger: PropTypes.func.isRequired,
  /**
   * Trigger mutations
   * @type {Function}
   */
  createDirectoryByPath: PropTypes.func,
  /**
   * Account Mutation, used to retrieve OAuth account
   * @type {Function}
   */
  findAccount: PropTypes.func,
  /**
   * Trigger mutation
   * @type {Function}
   */
  launchTrigger: PropTypes.func.isRequired,
  /**
   * Account mutation
   * @type {Func}
   */
  saveAccount: PropTypes.func.isRequired,
  /**
   * Account mutations
   * @type {Function}
   */
  watchKonnectorAccount: PropTypes.func.isRequired,
  /**
   * Trigger mutations
   * @type {Function}
   */
  statDirectoryByPath: PropTypes.func,
  /**
   * Job mutations
   * @type {Function}
   */
  watchKonnectorJob: PropTypes.func.isRequired,
  //
  // Callbacks
  //
  /**
   * Callback invoke when the konnector job fails
   * @type {Function}
   */
  onError: PropTypes.func,
  /**
   * Callback invoked when the trigger is launch
   * @type {Function}
   */
  onLaunch: PropTypes.func,
  /**
   * Callback invoked when the trigger has been launched and the login to the
   * remote service has succeeded.
   * @type {Function}
   */
  onLoginSuccess: PropTypes.func,
  /**
   * Callback invoked when the trigger has been launched and the job ended
   * successfully.
   * @type {Function}
   */
  onSuccess: PropTypes.func
}

export default withLocales(
  withMutations(
    accountsMutations,
    filesMutations,
    permissionsMutations,
    triggersMutations
  )(TriggerManager)
)
