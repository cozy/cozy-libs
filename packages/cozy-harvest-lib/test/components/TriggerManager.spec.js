/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'

import { DumbTriggerManager as TriggerManager } from 'components/TriggerManager'
import cronHelpers from 'helpers/cron'

jest.mock('cozy-doctypes', () => {
  const CozyFolder = {
    copyWithClient: () => CozyFolder,
    ensureMagicFolder: () => ({ path: '/Administrative' }),
    magicFolders: {
      ADMINISTRATIVE: 'io.cozy.apps/administrative',
      PHOTOS: '/photos'
    }
  }
  return {
    CozyFolder
  }
})

jest.mock('cozy-ui/transpiled/react/utils/color')

const fixtures = {
  data: {
    username: 'foo',
    passphrase: 'bar'
  },
  konnector: {
    slug: 'konnectest',
    fields: {
      username: {
        type: 'text'
      },
      passphrase: {
        type: 'password'
      }
    }
  },
  konnectorWithFolder: {
    name: 'myBills',
    slug: 'mybills',
    fields: {
      advancedFields: {
        folderPath: {
          advanced: true
        }
      }
    }
  },
  folder: {
    _id: '3f5b288af36041f189ec22063adab706'
  },
  folderPath: '/Administrative/myBills/foo',
  folderPermission: {
    saveFolder: {
      type: 'io.cozy.files',
      values: ['3f5b288af36041f189ec22063adab706'],
      verbs: ['GET', 'PATCH', 'POST']
    }
  },
  triggerAttributes: {
    arguments: '0 0 0 * * 0',
    type: '@cron',
    worker: 'konnector',
    message: {
      account: 'a87f9a8bd3884479a48811e7b7deec75',
      konnector: 'konnectest'
    }
  },
  account: {
    account_type: 'konnectest',
    auth: {
      username: 'foo',
      passphrase: 'bar'
    },
    identifier: 'username'
  },
  createdAccount: {
    _id: 'a87f9a8bd3884479a48811e7b7deec75',
    account_type: 'konnectest',
    auth: {
      username: 'foo',
      passphrase: 'bar'
    },
    identifier: 'username'
  },
  updatedAccount: {
    _id: 'a87f9a8bd3884479a48811e7b7deec75',
    account_type: 'konnectest',
    auth: {
      username: 'foo',
      passphrase: 'fuz'
    }
  },
  existingAccount: {
    _id: '61c683295560485db0f34b859197c581',
    account_type: 'konnectest',
    auth: {
      username: 'foo',
      passphrase: 'bar'
    },
    identifier: 'username'
  },
  existingTrigger: {
    id: '4b67e0bedd464704a6a995f5c2070ccf',
    _type: 'io.cozy.triggers',
    attributes: {
      arguments: '0 0 0 * * 0',
      type: '@cron',
      worker: 'konnector',
      message: {
        account: '61c683295560485db0f34b859197c581',
        konnector: 'konnectest'
      }
    }
  },
  createdTrigger: {
    id: '669e9a7cc3064a97bc0aa20feef71cb2',
    _type: 'io.cozy.triggers',
    attributes: {
      arguments: '0 0 0 * * 0',
      type: '@cron',
      worker: 'konnector',
      message: {
        account: 'a87f9a8bd3884479a48811e7b7deec75',
        konnector: 'konnectest'
      }
    }
  },
  launchedJob: {
    type: 'io.cozy.jobs',
    id: 'ac09e6f4473f4b6fbb83c9d2f532504e',
    domain: 'cozy.tools:8080',
    worker: 'konnector',
    state: 'running',
    queued_at: '2016-09-19T12:35:08Z',
    started_at: '2016-09-19T12:35:08Z',
    error: '',
    links: {
      self: '/jobs/ac09e6f4473f4b6fbb83c9d2f532504e'
    }
  },
  runningJob: {
    type: 'io.cozy.jobs',
    id: 'ac09e6f4473f4b6fbb83c9d2f532504e',
    domain: 'cozy.tools:8080',
    worker: 'konnector',
    state: 'running',
    queued_at: '2016-09-19T12:35:08Z',
    started_at: '2016-09-19T12:35:08Z',
    error: '',
    links: {
      self: '/jobs/ac09e6f4473f4b6fbb83c9d2f532504e'
    }
  },
  doneJob: {
    type: 'io.cozy.jobs',
    id: 'ac09e6f4473f4b6fbb83c9d2f532504e',
    domain: 'cozy.tools:8080',
    worker: 'konnector',
    state: 'done',
    queued_at: '2016-09-19T12:35:08Z',
    started_at: '2016-09-19T12:35:08Z',
    error: '',
    links: {
      self: '/jobs/ac09e6f4473f4b6fbb83c9d2f532504e'
    }
  }
}

const mockVaultClient = {
  createNewCipher: jest.fn(),
  saveCipher: jest.fn(),
  getByIdOrSearch: jest.fn(),
  decrypt: jest.fn(),
  createNewCozySharedCipher: jest.fn(),
  get: jest.fn(),
  shareWithCozy: jest.fn(),
  isLocked: jest.fn().mockResolvedValue(false)
}

const addPermissionMock = jest.fn()
const addReferencesToMock = jest.fn()
const createTriggerMock = jest.fn()
const createDirectoryByPathMock = jest.fn()
const statDirectoryByPathMock = jest.fn()
const launchTriggerMock = jest.fn()
const saveAccountMock = jest.fn()

const tMock = jest.fn()

const props = {
  addPermission: addPermissionMock,
  addReferencesTo: addReferencesToMock,
  konnector: fixtures.konnector,
  createTrigger: createTriggerMock,
  createDirectoryByPath: createDirectoryByPathMock,
  statDirectoryByPath: statDirectoryByPathMock,
  saveAccount: saveAccountMock,
  launch: launchTriggerMock,
  t: tMock,
  vaultClient: mockVaultClient,
  breakpoints: { isMobile: false }
}

const propsWithAccount = {
  ...props,
  account: fixtures.existingAccount,
  trigger: fixtures.existingTrigger
}

const shallowWithoutAccount = konnector =>
  shallow(
    <TriggerManager {...props} konnector={konnector || fixtures.konnector} />
  )

const shallowWithAccount = () =>
  shallow(<TriggerManager {...propsWithAccount} />)

describe('TriggerManager', () => {
  beforeEach(() => {
    createTriggerMock.mockResolvedValue(fixtures.createdTrigger)
    saveAccountMock.mockResolvedValue(fixtures.createdAccount)
    mockVaultClient.createNewCozySharedCipher.mockResolvedValue({
      id: 'cipher-id-1'
    })
  })

  afterEach(() => {
    jest.resetAllMocks()
  })

  it('should redirect to OAuthForm', () => {
    const konnector = {
      oauth: {
        scope: 'test'
      }
    }
    const component = shallow(
      <TriggerManager {...props} konnector={konnector} />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should render without account', () => {
    const component = shallowWithoutAccount().getElement()
    expect(component).toMatchSnapshot()
  })

  it('should render with account', () => {
    const component = shallowWithAccount().getElement()
    expect(component).toMatchSnapshot()
  })

  describe('handleError', () => {
    beforeEach(() => {
      statDirectoryByPathMock.mockResolvedValue(fixtures.folder)
      createDirectoryByPathMock.mockResolvedValue(fixtures.folder)
    })

    it('should render error', async () => {
      const wrapper = shallowWithAccount()
      await wrapper.instance().handleError(new Error('Test error'))
      expect(wrapper.getElement()).toMatchSnapshot()
    })

    const clientMutations = {
      saveAccount: saveAccountMock,
      createDirectoryByPath: createDirectoryByPathMock,
      addPermission: addPermissionMock,
      addReferencesTo: addReferencesToMock,
      createTrigger: createTriggerMock,
      statDirectoryByPath: statDirectoryByPathMock
    }

    for (var mutation of Object.keys(clientMutations)) {
      const mutationSync = mutation
      it(`should be called when ${mutation} fails`, async () => {
        clientMutations[mutationSync].mockRejectedValue(
          new Error(`${mutationSync} error`)
        )

        if (mutationSync !== 'statDirectoryByPath') {
          statDirectoryByPathMock.mockResolvedValue(null)
        }

        const wrapper = shallowWithoutAccount(fixtures.konnectorWithFolder)

        jest
          .spyOn(wrapper.instance(), 'handleError')
          .mockImplementation(() => {})

        await wrapper.instance().handleSubmit(fixtures.data)
        expect(wrapper.instance().handleError).toHaveBeenCalledWith(
          new Error(`${mutationSync} error`)
        )
      })
    }
  })

  describe('handleSubmit', () => {
    it('should render as submitting when there is no account', () => {
      const wrapper = shallowWithoutAccount()
      wrapper.instance().handleSubmit()
      expect(wrapper.state().status).toEqual('RUNNING') // TODO: test disabled prop of submit button instead of the internal state
    })

    it('should render as submitting when there is an account', () => {
      const wrapper = shallowWithAccount()
      wrapper.instance().handleSubmit()
      expect(wrapper.state().status).toEqual('RUNNING')
    })

    it('should call saveAccount without account', async () => {
      const wrapper = shallowWithoutAccount()
      await wrapper.instance().handleSubmit(fixtures.data)
      expect(saveAccountMock).toHaveBeenCalledWith(
        fixtures.konnector,
        expect.objectContaining(fixtures.account)
      )
    })

    it('should call saveAccount with account (no password provided)', async () => {
      const wrapper = shallowWithAccount()

      mockVaultClient.decrypt.mockResolvedValueOnce({
        login: {
          username: 'username',
          password: 'password'
        }
      })

      await wrapper.instance().handleSubmit({ login: 'test' })
      expect(saveAccountMock).toHaveBeenCalledWith(
        fixtures.konnector,
        expect.objectContaining(fixtures.existingAccount)
      )
    })

    it('should call saveAccount with account with RESET_SESSION state if passphrase changed', async () => {
      const wrapper = shallowWithAccount()

      mockVaultClient.decrypt.mockResolvedValueOnce({
        login: {
          username: 'username',
          password: 'password'
        }
      })

      await wrapper.instance().handleSubmit(fixtures.data)
      expect(saveAccountMock).toHaveBeenCalled()
      expect(saveAccountMock.mock.calls[0][1].state).toEqual('RESET_SESSION')
    })

    it('should call saveAccount with account with RESET_SESSION state if password changed', async () => {
      const wrapper = shallowWithAccount()
      const instance = wrapper.instance()

      mockVaultClient.decrypt.mockResolvedValueOnce({
        login: {
          username: 'username',
          password: 'password'
        }
      })

      await instance.handleSubmit({
        login: 'foo',
        password: 'bar'
      })

      expect(saveAccountMock).toHaveBeenCalled()
      expect(saveAccountMock.mock.calls[0][1].state).toEqual('RESET_SESSION')
    })

    it('should call handleNewAccount with account', async () => {
      const wrapper = shallowWithAccount()
      const instance = wrapper.instance()
      jest
        .spyOn(instance, 'handleNewAccount')
        .mockResolvedValue(fixtures.launchedJob)

      mockVaultClient.decrypt.mockResolvedValueOnce({
        login: {
          username: 'username',
          password: 'password'
        }
      })

      await instance.handleSubmit(fixtures.data)

      expect(instance.handleNewAccount).toHaveBeenCalledWith(
        fixtures.createdAccount
      )
    })

    it('should call handleNewAccount without account', async () => {
      const wrapper = shallowWithoutAccount()
      const instance = wrapper.instance()
      jest
        .spyOn(instance, 'handleNewAccount')
        .mockResolvedValue(fixtures.launchedJob)

      await instance.handleSubmit(fixtures.data)

      expect(instance.handleNewAccount).toHaveBeenCalledWith(
        fixtures.createdAccount
      )
    })
  })

  describe('handleNewAccount', () => {
    beforeAll(() => {
      jest.spyOn(cronHelpers, 'fromFrequency').mockReturnValue('0 0 0 * * 0')
    })

    afterEach(() => {
      jest.clearAllMocks()
    })

    afterAll(() => {
      jest.restoreAllMocks()
    })

    it('should create trigger', async () => {
      const wrapper = shallowWithoutAccount()
      await wrapper.instance().handleNewAccount(fixtures.createdAccount)
      expect(createTriggerMock).toHaveBeenCalledTimes(1)
      expect(createTriggerMock).toHaveBeenCalledWith(fixtures.triggerAttributes)
    })

    it('should not create trigger when one is passed as prop', async () => {
      const wrapper = shallow(<TriggerManager {...propsWithAccount} />)
      await wrapper.instance().handleNewAccount(fixtures.updatedAccount)
      expect(createTriggerMock).not.toHaveBeenCalled()
    })

    it('should launch trigger without account', async () => {
      const wrapper = shallowWithoutAccount()
      await wrapper.instance().handleNewAccount(fixtures.account)
      expect(launchTriggerMock).toHaveBeenCalledTimes(1)
      expect(launchTriggerMock).toHaveBeenCalledWith(fixtures.createdTrigger)
    })

    describe('when konnector needs folder', () => {
      beforeEach(() => {
        tMock.mockReturnValue('/Administrative')
      })

      afterEach(() => {
        tMock.mockReset()
      })

      it('should create folder if it does not exist', async () => {
        statDirectoryByPathMock.mockResolvedValue(null)
        createDirectoryByPathMock.mockReturnValue(fixtures.folder)

        const wrapper = shallowWithoutAccount(fixtures.konnectorWithFolder)
        await wrapper.instance().handleNewAccount(fixtures.account)

        expect(statDirectoryByPathMock).toHaveBeenCalledTimes(1)
        expect(createDirectoryByPathMock).toHaveBeenCalledTimes(1)
        expect(createDirectoryByPathMock).toHaveBeenCalledWith(
          fixtures.folderPath
        )
        expect(addPermissionMock).toHaveBeenCalledTimes(1)
        expect(addPermissionMock).toHaveBeenCalledWith(
          fixtures.konnectorWithFolder,
          fixtures.folderPermission
        )
        expect(addReferencesToMock).toHaveBeenCalledTimes(1)
        expect(addReferencesToMock).toHaveBeenCalledWith(
          fixtures.konnectorWithFolder,
          [fixtures.folder]
        )
      })

      it('should not create folder if it exists', async () => {
        statDirectoryByPathMock.mockResolvedValue(fixtures.folder)

        const wrapper = shallowWithoutAccount(fixtures.konnectorWithFolder)
        await wrapper.instance().handleNewAccount(fixtures.account)

        expect(statDirectoryByPathMock).toHaveBeenCalledTimes(1)
        expect(createDirectoryByPathMock).toHaveBeenCalledTimes(0)

        expect(addPermissionMock).toHaveBeenCalledTimes(1)
        expect(addPermissionMock).toHaveBeenCalledWith(
          fixtures.konnectorWithFolder,
          fixtures.folderPermission
        )
        expect(addReferencesToMock).toHaveBeenCalledTimes(1)
        expect(addReferencesToMock).toHaveBeenCalledWith(
          fixtures.konnectorWithFolder,
          [fixtures.folder]
        )
      })
    })

    it('should launch trigger with account', async () => {
      const wrapper = shallowWithAccount()
      await wrapper.instance().handleNewAccount(fixtures.updatedAccount)
      expect(launchTriggerMock).toHaveBeenCalledTimes(1)
      expect(launchTriggerMock).toHaveBeenCalledWith(fixtures.existingTrigger)
    })

    it('should keep updated account in state', async () => {
      const wrapper = shallowWithAccount()
      await wrapper.instance().handleNewAccount(fixtures.updatedAccount)
      expect(wrapper.state().account).toEqual(fixtures.updatedAccount)
    })
  })
})
