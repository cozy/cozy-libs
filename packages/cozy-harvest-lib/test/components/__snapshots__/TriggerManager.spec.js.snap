// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TriggerManager handleError should render error 1`] = `
<div>
  <Wrapper>
    <div
      id="coz-harvest-modal-place"
    />
    <React.Fragment>
      <Wrapper
        account={
          Object {
            "_id": "61c683295560485db0f34b859197c581",
            "account_type": "konnectest",
            "auth": Object {
              "passphrase": "bar",
              "username": "foo",
            },
            "identifier": "username",
          }
        }
        error={[Error: Test error]}
        konnector={
          Object {
            "fields": Object {
              "passphrase": Object {
                "type": "password",
              },
              "username": Object {
                "type": "text",
              },
            },
            "slug": "konnectest",
          }
        }
        onBack={[Function]}
        onSubmit={[Function]}
        submitting={false}
      />
    </React.Fragment>
  </Wrapper>
</div>
`;

exports[`TriggerManager should redirect to OAuthForm 1`] = `
<Wrapper
  konnector={
    Object {
      "oauth": Object {
        "scope": "test",
      },
    }
  }
  onSuccess={[Function]}
  submitting={false}
/>
`;

exports[`TriggerManager should render with account 1`] = `
<div>
  <Wrapper>
    <div
      id="coz-harvest-modal-place"
    />
    <React.Fragment>
      <Wrapper
        account={
          Object {
            "_id": "61c683295560485db0f34b859197c581",
            "account_type": "konnectest",
            "auth": Object {
              "passphrase": "bar",
              "username": "foo",
            },
            "identifier": "username",
          }
        }
        konnector={
          Object {
            "fields": Object {
              "passphrase": Object {
                "type": "password",
              },
              "username": Object {
                "type": "text",
              },
            },
            "slug": "konnectest",
          }
        }
        onBack={[Function]}
        onSubmit={[Function]}
        submitting={false}
      />
    </React.Fragment>
  </Wrapper>
</div>
`;

exports[`TriggerManager should render without account 1`] = `
<div>
  <Wrapper>
    <div
      id="coz-harvest-modal-place"
    />
    <Wrapper
      konnector={
        Object {
          "fields": Object {
            "passphrase": Object {
              "type": "password",
            },
            "username": Object {
              "type": "text",
            },
          },
          "slug": "konnectest",
        }
      }
      onNoCiphers={[Function]}
      onSelect={[Function]}
    />
  </Wrapper>
</div>
`;
